src/Pool.sol

```diff
@@ -231,46 +230,52 @@ abstract contract Pool {
     /// @return withdrawn The withdrawn amount which should be sent to the user.
     /// Equal to `withdrawAmt` unless `WITHDRAW_ALL` is used.
     function _updateSenderInternal(
-        address id,
+        address senderAddr,
         uint128 topUpAmt,
         uint128 withdrawAmt,
         uint128 amtPerSec,
         ReceiverWeight[] calldata updatedReceivers
     ) internal returns (uint128 withdrawn) {
-        uint256 maxUpdates = senders[id].weightCount + updatedReceivers.length;
+        Sender storage sender = senders[senderAddr];
```

Comment:
If gas is really an issue here ,we could either us a global sender mapping for sender and subsenders where the hash(address) or the hash(address, subsenderId) is the key.

The other approach would be to maybe pass to `_toUp`, `_stopSending`, etc methods the subSenderId and indicate with a flag from which mapping it should be read.

Reply:
You're right, the gas usage IS worse after switching to passing by reference, usually by 1-2K gas per test case. I wish I understood why, to me it doesn't make sense. It should be cheaper, because there's no hashing of the keys on each access.

The other factor is events emission postponing, the reduced number of emitted events doesn't fully recoup the memory storage and iteration overhead.

I'm very tempted to accept the higher gas usage because:
- It's relatively small, the gas usage of the pool is already obscene and we need bold optimizations anyway.
- Passing around a sender reference is a good start to optimization. E.g. it's easy to replace with a version cached in memory or with calldata when (if) we move the sender data out of the storage.
- It's a relatively simple change and it's working. We don't duplicate code and keep the interfaces clean by abstracting sender type.
- At this point I'm doubting common sense when it comes to deciding what will be cheaper. E.g. putting both sender types into a single mapping by hashing them differently WILL introduce complexity and WILL take time, but it only MAY reduce gas usage. I really don't know if it'll shave these additional 1-2K gas introduced by this PR or if it'll add even more.

What do you think?

Reply:
Is it not way more gas?
```
> [PASS] testAllowsToppingUpWhileSending() (gas: 411369)
vs.
> [PASS] testAllowsToppingUpWhileSending() (gas: 660886)
```
I don't know how passing by reference is implemented internally. We could also maybe try to do it with inline assembly and SLOAD in the helper functions and passing the storage pointer?

I was curious and wanted to see how much it would be if we read it into memory but thats not possible because of the nested mapping.

```
TypeError: Type struct Pool.Sender is only valid in storage because it contains a (nested) mapping.
```

Reply:
Yeah, Solidity dropped silent dropping non-memory fields of structs. Now we'd need to split up sender into the common part and the receivers, but that's a different story.

Huh, I'm getting totally different gas usage. Are you sure that you're running the right thing?
```
> git rev-parse HEAD
13951fafde01c2f86dedd961f2f2c218bc3b38f1
> make test
...
Running 25 tests for src/test/ERC20Pool.t.sol:ERC20PoolTest
[PASS] testAllowsToppingUpWhileSending() (gas: 411718)
...
Running 25 tests for src/test/EthPool.t.sol:EthPoolTest
[PASS] testAllowsToppingUpWhileSending() (gas: 328568)
```

Reply:
I compared the master branch with the CI.

Locally, I am getting
```
[PASS] testAllowsToppingUpWhileSending() (gas: 411718)
``` 
same as us. I am using the following dapp tools version
```
dapp 0.34.1
solc, the solidity compiler commandline interface
Version: 0.8.7+commit.e28d00a7.Darwin.appleclang
hevm 0.48.1
```